WebView2: Rewrote Distribution and Versioning articles (#1331)
MicrosoftDocs/edge-developer: microsoft-edge/webview2/concepts/versioning.md

title: Understand WebView2 SDK versions
author: MSEdgeTeam
ms.author: msedgedevrel
ms.date: 07/13/2021
ms.topic: conceptual
ms.prod: microsoft-edge
ms.technology: webview
keywords: IWebView2, IWebView2WebView, webview2, webview, wpf apps, wpf, edge, ICoreWebView2, ICoreWebView2Host, browser control, edge html
---
# Understand WebView2 SDK versions

New versions of the WebView2 SDK are shipped at the same general cadence as the Microsoft Edge \(Chromium\) browser, which is approximately every six weeks.  

When you develop and distribute a WebView2 app, the client must at least have a version of the Runtime that can use the core APIs and create a WebView2.  The client ideally should have a version of the Runtime that supports all of the latest APIs that are in the SDK version you use to develop the app.  You must coordinate the WebView2 SDK version that you use for development, and the WebView2 Runtime version that's installed on client machines.

> *  Core API functionality requires that the client has WebView2 Runtime version `86.0.616.0` or newer.

Reviewer: I would avoid using the word Core here and in other places in this doc, as it may confuse folks to think this refers to anything in the Core namespace (https://docs.microsoft.com/en-us/dotnet/api/microsoft.web.webview2.core?view=webview2-dotnet-1.0.864.35), which is not true.

I think what we want to convey in this doc is that `86.0.616.0` is the necessary (but not sufficient) min version for any WebView2 API. WebView2 APIs are forward-compatible and can work on a Runtime that has a higher build number than the SDK build number in which the API is introduced. E.g. API_x is introduced in SDK 1.0.900.0, so it would work in runtime 94.0.1000.0 but not runtime 90.0.700.0. I'm making numbers up here but hopefully you get the point. `86.0.616.0` is kinda special because this is the version we GA with, so any WebView2 API that we GA'ed with would work on 616 and onwards. What you mentioned as Core is basically the set of API we GA'ed with. 

When developers are using an SDK that has lower or equal build number than runtime, every API they have access to works. When developers are using an SDK that has a greater build number than runtime, some of the APIs' implementation are not yet available in the runtime therefore cannot work. Since evergreen developers can't really control the runtime version (it pausing updates or offline), it's a good practice to feature-detect (QI/try-catch) when using relatively new APIs.

*  Full support for the latest APIs in an SDK release requires that the Runtime on the client have a version number greater than or equal to the SDK version.

## Release and prerelease packages

The NuGet package for the WebView2 SDK contains both a release and prerelease package.

*  The _release_ package of the SDK contains all of the stable Win32 C/C++ and .NET APIs.  APIs that are in the release package are fully supported.  The release package is forward-compatible and contains the following components:
    *  [Win32 C/C++ APIs][ReferenceWin32].
    *  .NET APIs:  [WPF][DotnetMicrosoftWebWebview2WpfNamespace], [WinForms][DotnetMicrosoftWebWebview2WinformsNamespace], and [Core][DotnetMicrosoftWebWebview2CoreNamespace].  

*  The _prerelease_ package of the SDK is a superset of the release package, with additional, [Experimental APIs](#experimental-apis) that are subject to change based on your feedback.  Avoid using the prerelease version of the WebView2 SDK to build production apps.

> ## Experimental APIs  

Reviewer: I would move the Experimental API stuff you had in the distribution doc (see `### Experimental APIs are not guaranteed to be forward-compatible`) to here.

To manually look up the Runtime or browser version requirements for full API support for a specific SDK version, navigate to [Release notes for WebView2 SDK](../release-notes.md).  Then check the section for the specific SDK release.

In your WebView2 app, to programmatically test whether a recent API is supported, a best practice is to check for support by using `QueryInterface` or `try/catch`.  This approach is described below.

## Testing against non-stable channels
   
> [!IMPORTANT]
> When developing [Evergreen WebView2 apps][Webview2ConceptsDistributionEvergreenDistributionMode], regularly test your app against the latest non-stable Microsoft Edge channels and the WebView2 Runtime.  Because the web platform is constantly evolving, regular testing is the best way to ensure your app performs as intended.

## Testing whether the installed Runtime supports a recent API

When you develop a WebView2 app using a recent version of the WebView2 SDK, when you use a recently added API, it's a good practice to test the client's installed WebView2 Runtime to check whether that API is present.  How your app programmatically tests for API support depends on the coding platform.

> *   **Win32 C/C++**.  When using `QueryInterface` to obtain a new interface, test for a return value of `E_NOINTERFACE`.  That return value likely indicates that the client's WebView2 Runtime is an older version that doesn't support that interface.  For an example of checking for the existence of specific WebView2 APIs in the Runtime, navigate to [Line 622 - AppWindow.cpp][GithubMicrosoftedgeWebview2samplesSampleappsWebview2apisampleAppwindowCppL622].<!-- outdated?  that cpp file contains neither "catch" nor "QueryInterface" nor "E_NOINTERFACE" -->

Author: @liminzhu - "navigate to Line 622 of AppWindow.cpp" - Is this statement outdated?  That cpp file contains neither "catch" nor "QueryInterface" nor "E_NOINTERFACE".

Reviewer: yah we update our sample app pretty regularly, so the lines have changed.

Author: Added mention of CHECK_FAILURE function & CheckFailure.h.

*   **.NET and WinUI**.  Use `try/catch` and check for a `No such interface supported` exception when using methods, properties, and events that were added to more recent versions of the WebView2 SDK.  This exception likely indicates that the client's WebView2 Runtime is an older version that doesn't support that API.
    

> If an API is unavailable in the client's installed WebView2 Runtime, consider turning off the associated feature for the user, or inform the user that they must update the WebView2 Runtime to use the feature.

Reviewer: ```suggestion
If an API is unavailable in the client's installed WebView2 Runtime, consider providing graceful fallback for the associated feature for the user, or inform the user that they must update the WebView2 Runtime to use the feature.
```

<!-- Is this section commented out because deprecated?  delete or uncomment the section?

> ## Versioning  

Author: Is this section commented out because deprecated? delete or uncomment the section?

Author: I rewrote but kept section commented-out.

To use the WebView2 APIs, 86.0.616.0 is the necessary (but not sufficient) minimum version of Microsoft Edge or the WebView2 Runtime.
WebView2 APIs are forward-compatible; a WebView2 API works when using a WebView2 Runtime that has an equal or higher build number as the SDK build number in which the API was introduced.

For example, if an API is introduced in SDK 1.0.900.0, that API would work with Runtime 94.0.1000.0, but not with Runtime 90.0.700.0.  The build number is the third part of the four-part version number for the Webview2 SDK and for Microsoft Edge or the WebView2 Runtime.

Runtime 86.0.616.0 is a special release, because it is the General Availability release.  Any WebView2 API that was in the General Availability release (WebView2 SDK version 616) works with Runtime version 616 and later.

When you use a WebView2 SDK that has a build number equal to or less than the WebView2 Runtime, every API that you have access to in that SDK works with that version of the Runtime.  When you use a WebView2 SDK that has a build number greater than the WebView2 Runtime, the newer APIs' implementations are not yet available in the Runtime, and therefore cannot work.

> If you use the Evergreen Runtime, you can't guarantee that a client's Runtime has been updated to be equal to or greater than your SDK build number, because the Admin might have turned off updating of the WebView2 Runtime, or the client might be offline long-term.  Therefore, it's a good practice to wrap any calls to relatively new WebView2 APIs by using `QueryInterface` or `try-catch`.

Reviewer: ```suggestion
If you use the Evergreen Runtime, there are edge cases where a client's Runtime is not the latest and therefore falls behind the SDK build number, because the Admin might have turned off updating of the WebView2 Runtime, or the client might be offline long-term.  Therefore, it's a good practice to wrap any calls to relatively new WebView2 APIs by using `QueryInterface` or `try-catch`.
```

Making it sounds scary.

<!-- todo: dissolve/merge section with the above -->
### Details

When you develop and distribute a WebView2 app, the client must at least have a version of the Runtime that can use the set of APIs in the WebView2 General Availability release (build 616) and create a WebView2.  The client ideally should have a version of the Runtime that supports all of the latest APIs that are in the SDK version you use to develop the app.  You must coordinate the WebView2 SDK version that you use for development, and the WebView2 Runtime version that's installed on client machines.

> *  The set of APIs in the WebView2 General Availability release (build 616) requires that the client has WebView2 Runtime version `86.0.616.0` or newer.

Reviewer: ```suggestion
*  The set of APIs in the WebView2 General Availability release (build 616) requires that the client has WebView2 Runtime version `86.0.616.0` or higher.
```

Author: fixed throughout this file.

> *  Full support for the latest APIs in an SDK release requires that the Runtime on the client have a version number greater than or equal to the SDK version.

Reviewer: ```suggestion
*  Full support for the latest APIs in an SDK release requires that the Runtime on the client have a build number greater than or equal to the SDK version.
```

## Release and prerelease packages

The NuGet package for the WebView2 SDK contains both a release and prerelease package.

*  The _release_ package of the SDK contains all of the stable Win32 C/C++ and .NET APIs.  APIs that are in the release package are fully supported.  The release package is forward-compatible and contains the following components:
    *  [Win32 C/C++ APIs][ReferenceWin32].
    *  .NET APIs:  [WPF][DotnetMicrosoftWebWebview2WpfNamespace], [WinForms][DotnetMicrosoftWebWebview2WinformsNamespace], and [Core][DotnetMicrosoftWebWebview2CoreNamespace].  

*  The _prerelease_ package of the SDK is a superset of the release package, with additional, [Experimental APIs](#experimental-apis) that are subject to change based on your feedback.  Avoid using the prerelease version of the WebView2 SDK to build production apps.

<!-- possibly condense section, having added para3 -->
## Experimental APIs  

<!-- from paragraph 3 of post from github issue 341: -->
### Experimental APIs are not guaranteed to be forward-compatible

> Any experimental APIs in a pre-release SDK are not guaranteed to be forward-compatible.   When a _pre-release_ version of the WebView2 SDK is initially made available, that SDK only works with Microsoft Edge Canary.  After that, the SDK works with the Beta and Dev channels, for the development phase of your app.  A pre-release version of the WebView2 SDK is meant for developers to try out new APIs early and provide feedback before the new APIs are promoted to become stable, forward-compatible APIs.

Reviewer: ```suggestion
Any experimental APIs in a pre-release SDK are not guaranteed to be forward-compatible and the support may be completely removed in future Runtime updates.   When a _pre-release_ version of the WebView2 SDK is initially made available, that SDK only works with Microsoft Edge Canary.  After that, the SDK works with the Beta and Dev channels, for the development phase of your app.  A pre-release version of the WebView2 SDK is meant for developers to try out new APIs early and provide feedback before the new APIs are promoted to become stable, forward-compatible APIs.
```

For a release version of your app, use a _Release_ version of the WebView2 SDK.  The _stable_ (non-experimental) APIs that are in that SDK version are forward-compatible.  

> The WebView2 SDK has been forward-compatible ever since version 1.  If your WebView2 app uses the current version of the SDK, your app continues to work even after the client's WebView2 Runtime is updated.  However, any experimental APIs that are in a version of the WebView2 SDK are not guaranteed to be forward-compatible.

Reviewer: ```suggestion
The WebView2 _release_ SDK has been forward-compatible ever since version 1.  If your WebView2 app uses the current version of the SDK, your app continues to work even after the client's WebView2 Runtime is updated.  However, any experimental APIs that are in a version of the WebView2 SDK are not guaranteed to be forward-compatible.
```

> [!NOTE]
> Experimental APIs might not be available in your installed version of the WebView2 Runtime.  

## Matching the Runtime version with the SDK version

A WebView2 production app requires that the [WebView2 Runtime][MicrosoftDeveloperEdgeWebview2] is present on the client.  In the Evergreen distribution approach, the installed WebView2 Runtime automatically updates to the latest version available; however, a user or IT admin might choose to prevent automatic updating of the WebView2 Runtime.  The resulting outdated Runtime on the client might cause compatibility issues between your updated WebView2 app that uses new APIs from a recent SDK, and the older Runtime.

In case updating the WebView2 Runtime is prevented on the client, make sure that you know the minimum version number of the [WebView2 Runtime][MicrosoftDeveloperEdgeWebview2] that is required by your app.  The minimum required Runtime version to support the General Availability release of the SDK (build 616) is older than for the latest Runtime.  The latest Runtime supports all APIs that are in the latest SDK release build.

> ### Minimum version for GA API support

Reviewer: For  `### Minimum version for GA API support` and `### Minimum version for full API support`, didn't we talk about the same things in similar details in the first few paragraphs of the doc?

In your WebView2 app, to programmatically test whether a recent API is supported, a best practice is to check for support by using `QueryInterface` or `try/catch`.  This approach is described below.

## Testing against Microsoft Edge preview channels
   
> [!IMPORTANT]
> When developing Evergreen WebView2 apps, regularly test your app against the latest Microsoft Edge preview channels and the WebView2 Runtime.  Because the web platform is constantly evolving, regular testing is the best way to ensure your app performs as intended.

For information about Evergreen WebView2 Runtime distribution, navigate to [Distribute a WebView2 app and the WebView2 Runtime][Webview2ConceptsDistribution].

## Testing whether the installed Runtime supports a recent API

When you develop a WebView2 app using a recent version of the WebView2 SDK, when you use a recently added API, it's a good practice to test the client's installed WebView2 Runtime to check whether that API is present.  How your app programmatically tests for API support depends on the coding platform.

> *   **Win32 C/C++**.  When using `QueryInterface` to obtain a new interface, test for a return value of `E_NOINTERFACE`.  That return value likely indicates that the client's WebView2 Runtime is an older version that doesn't support that interface.  For an example of checking for the existence of specific WebView2 APIs in the Runtime, navigate to [Line 622 - AppWindow.cpp][GithubMicrosoftedgeWebview2samplesSampleappsWebview2apisampleAppwindowCppL622].  This file wraps WebView2 API calls in the `CHECK_FAILURE` macro function, defined in `CheckFailure.h`.

Reviewer: > Line 622 - AppWindow.cpp

Let's advise folks to search for `try_query` on AppWindow.cpp to look for examples instead of giving out a line number since that number will end up changing all the time as we update the sample.

author: MSEdgeTeam
ms.author: msedgedevrel
ms.date: 08/03/2021
ms.topic: conceptual
ms.prod: microsoft-edge
ms.technology: webview
keywords: IWebView2, IWebView2WebView, webview2, webview, wpf apps, wpf, edge, ICoreWebView2, ICoreWebView2Host, browser control, edge html
---
# Understand WebView2 SDK versions

<!-- todo: cover Fixed Version.  when/if to specify "the Evergreen Runtime", instead of just "the Runtime".  does it work, contrasting "preview channel" vs. "Runtime" as terminology for mutually exclusive options? -->

The NuGet package for the WebView2 SDK contains both a release and prerelease package.  Either use a prerelease SDK with a preview channel of Microsoft Edge, or use a release SDK with the WebView2 Runtime.

*   During development and testing of your WebView2 app, to use experimental WebView2 APIs, use the following combination:
    *   A _prerelease_ version of the WebView2 SDK.
    *   A _preview channel_ of Microsoft Edge.

*   For a production release of your WebView2 app, use the following combination:
    *   A _release_ version of the WebView2 SDK.
    *   The WebView2 _Runtime_.

These two combinations are described below.

> ## Use a prerelease version of the SDK along with a preview channel of Microsoft Edge

Reviewer: This part might be slightly misleading. It sounds like when you're developing, you use pre-release + non-stable; when you're in prod, you use release + runtime. This isn't the case.

We have pre-release and release packages. Pre-release packages are for developers during development who want to test our latest WebView2 APIs, including the experimental APIs, before its support land in the Runtime. Developers using pre-release packages are recommended to use it against Canary so that the APIs are actually implemented here. 

Release packages only contain stable API. Developers can use them either in development (because not all developers care about the latest experimental APIs that we expose), or in production.
